### PR TITLE
Add article comparing Elixir and Clojure code

### PR DESCRIPTION
**Pull request by michaelstalker**
I also fixed some grammar in the Elixir/Clojure overview article.

### CODE REVIEW

File: articles/clojure_elixir_fibonacci_mstalker/article.md [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+### Elixir
+
+How does Elixir handle the simple implementation? Here is how we could rewrite the Clojure code in Elixir:
+
+```elixir
+defmodule Fibonacci do
+  def compute(n) do
+    cond do
+      n == 0 -> 0
```
Comment (doomspork) [inline articles/clojure_elixir_fibonacci_mstalker/article.md:71]: If you're going to compare the examples I think it's worth making the code equivalent:

```elixir
defmodule Fibonacci do
  def compute(n) do
    cond do
      n < 2 -> n
      true -> compute(n - 1) + compute(n - 2)
    end
  end
end
```

Comment (michaelstalker) [reply articles/clojure_elixir_fibonacci_mstalker/article.md:71]: I see what you mean. My goal was to write an implementation in each language using both 0 and 1 explicitly, and then to write an implementation where I consolidate the 0 and 1 cases using an inequality or a guard clause. Maybe I'll just take the inequality and guard clause cases out. Including the Clojure inequality doesn't really enhance the article, and omitting the Elixir `n in [0, 1]` won't hurt, either. I use a guard clause later in the article. What do you think?

Comment (michaelstalker) [reply articles/clojure_elixir_fibonacci_mstalker/article.md:71]: If I remove some code samples, do you think it would be better to eliminate this version:
```clojure
(defn compute
  [n]
  (cond
    (= n 0) 0
    (= n 1) 1
    :else (+ (compute (- n 1))
             (compute (- n 2)))))
```
or this version?
```clojure
(defn compute
  [n]
  (cond
    (< n 2) n
    :else (+ (compute (- n 1))
             (compute (- n 2)))))
```

File: articles/clojure_elixir_fibonacci_mstalker/article.md [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+We can call our function like this:
+
+```clojure
+(fibonacci/compute 10)
+;; 55
+```
+
+A semicolon starts a comment in Clojure. I wrote the return value of our function call after `;;`. Using two semicolons is idiomatic, and the language would have allowed us to use just one.
+
```
Comment (mpelzsherman) [inline articles/clojure_elixir_fibonacci_mstalker/article.md]: I found the above a bit confusing. If you want to demonstrate comments, I'd do it in the context of your function code. Maybe show the output of the function call in a terminal screen shot or something. Better yet, offer a "try it!" feature where people can interactively call the function in an embedded REPL.

File: articles/clojure_elixir_fibonacci_mstalker/article.md (hunk 1 above)
Comment (mpelzsherman) [reply articles/clojure_elixir_fibonacci_mstalker/article.md:71]: For the purpose of this blog post, I think it's a bit clearer to write 0 and 1 as explicit cases.

File: articles/clojure_elixir_fibonacci_mstalker/article.md [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+      n == 1 -> 1
+      true -> compute(n - 1) + compute(n - 2)
+    end
+  end
+end
+```
+
+First, I defined a module with `defmodule`. Modules are like Clojure's namespaces, and they contain zero or more functions. Then, I used `def` to define a function named `compute/1`. This function takes argument, `n`.
+
+Did you notice that I called this function `compute/1` instead of `compute`? We use both the function name and the number of arguments when referring to functions in Elixir.
```
Comment (mpelzsherman) [inline articles/clojure_elixir_fibonacci_mstalker/article.md:81]: Does Elixir support [function overloading](https://en.wikipedia.org/wiki/Function_overloading)? Does Clojure?
Probably worth mentioning that here.

File: articles/clojure_elixir_fibonacci_mstalker/article.md [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+end
+```
+
+A guard clause adds a condition to the function clauses. If the pattern of the argument matches the function signature, and the arguments satisfy the guard clause, the function executes; otherwise, it doesn't. In this case, I used a membership guard called [`in`](https://hexdocs.pm/elixir/Kernel.html#in/2). If `n` is equal to any element in the list `[0, 1]`, Elixir executes the first function clause and returns `n`. Otherwise, Elixir executes the second function clause.
+
+
+## Optimizing Our Functions
+
+The functions above are fine for small values of `n`. If `n` is sufficiently large, though, these functions will consume a large quantity of memory and will slow down significantly. The Clojure version even raises a `StackOverflow Error`. Try calling these functions when `n` is `200000` and you'll see what I mean!
+
```
Comment (mpelzsherman) [inline articles/clojure_elixir_fibonacci_mstalker/article.md:117]: Suggest changing to "Try calling these functions with `n` = `200000` ".

File: articles/clojure_elixir_fibonacci_mstalker/article.md [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+  (cond (= n 0) 0
+        (= n 1) 1
+        :else (helper n)))
+```
+
+The `compute` function looks very similar to our original implementation. We return `0` or `1` if `n` is `0` or `1`, respectively. Otherwise, we execute another function to calculate the value we want. In this case we use a function called `helper`.
+
+`defn-` is a private function. Due to how Clojure compiles code, we have to define `helper` above any code that uses it.
+
+Clojure isn't tail-call-optimized, but it has a [`loop`](http://clojuredocs.org/clojure.core/loop)/[`recur`](http://clojuredocs.org/clojure.core/recur) construct that works around this. In `helper`, we set up a starting point for our recursion with `loop`, and we initialize some values. Our assumption is that if you're in this function, the index you're looking for is at least `2`, so we give `current-index` an initial value of `2`. `current-value` is `1` when the index is `2`. The previous Fibonacci value is `1`, and we call this `previous-value` in our loop.
```
Comment (mpelzsherman) [inline articles/clojure_elixir_fibonacci_mstalker/article.md]: I'm not sure most readers will be familiar with tail-call optimization. Maybe add a reference?

File: articles/clojure_elixir_fibonacci_mstalker/article.md [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+(ns fibonacci)
+
+(defn- helper
+  [n]
+  (loop [current-index 2
+         current-value 1
+         previous-value 1]
+    (if (= n current-index)
+      current-value
+      (recur (inc current-index)
```
Comment (mpelzsherman) [inline articles/clojure_elixir_fibonacci_mstalker/article.md:135]: Is it necessary (or somehow better) to use `inc` to increment a number in clojure?
Why not just `(+ current-index 1)` ?

File: articles/clojure_elixir_fibonacci_mstalker/article.md [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+
+(defn compute
+  [n]
+  (let [fib (map first (iterate (fn [[a b]] [b (+' a b)]) [0 1]))]
+    (nth fib n)))
+```
+
+Let's interpret this from the inside out. `(fn [[a b]] [b (+' a b)])` is an anonymous function. It takes a two-element [vector](https://clojure.org/guides/learn/sequential_colls#_vectors) as its argument, and binds the first and second elements to `a` and `b`, respectively. It returns a two-element vector. The first element is `b`, and the second element is the sum of `a` and `b`.
+
+`iterate` generates a lazy sequence. The first element is `[0 1]`. `iterate` passes `[0, 1]` to the anonymous function to generate the second element, `[1 1]`. `iterate` generates the third sequence element by passing the second element, `[1 1]`, to the anonymous function, and so on.
```
Comment (mpelzsherman) [inline articles/clojure_elixir_fibonacci_mstalker/article.md]: Insert a reference to further reading on lazy sequences here?

**Comment by mmaiza** [comment]
Looks pretty good, but maybe you should add some guard against negative numbers

File: articles/clojure_elixir_mstalker/article.md [hunk 8]
```diff
@@ -21,7 +21,7 @@ rather than
 1 + 2
 ```
 
-This takes a bit of getting used to. However, once you do, it affords you some nice advantages. The structure of your code represents the abstract syntax tree (AST) very closely. Moreover, some editing tools like [ParEdit](https://www.emacswiki.org/emacs/ParEdit) or [Smartparens](https://github.com/Fuco1/smartparens) let you manipulate your source code very quickly. Clojure also lets you pass a variable number of arguments functions like `+`:
+This takes a bit of getting used to. However, once you do, it affords you some nice advantages. The structure of your code represents the abstract syntax tree (AST) very closely. Moreover, some editing tools like [ParEdit](https://www.emacswiki.org/emacs/ParEdit) or [Smartparens](https://github.com/Fuco1/smartparens) let you maniuplate your source code very quickly. Clojure also lets you pass a variable number of arguments to functions like `+`:
```
Comment (Stratus3D) [inline articles/clojure_elixir_mstalker/article.md]: I think this is a combination of new and old changes. `maniuplate` needs to be reverted.

File: articles/clojure_elixir_fibonacci_mstalker/article.md [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+
+The first two elements are `0` and `1`. After that, each element is the sum of the previous two elements.
+
+Here's today's programming challenge:
+
+> For any non-negative integer `n`, write a function to retrieve the `n`th Fibonacci number. Use a zero-based index.
+
+For example, when `n` is `4`, our function should return `3`.
+
+My goal in the following code exploration is to highlight some language features of Clojure and Elixir. I'll give examples of how someone might solve our programming challenge in either language, and then I'll explain some syntax or conventions. I'm not attempting to produce the most efficient or the cleverest code, as that may actually get in the way of giving you a taste for each language!
```
Comment (Stratus3D) [inline articles/clojure_elixir_fibonacci_mstalker/article.md:21]: The most common Fibonacci algorithms in Elixir are very slow. I wrote a blog post about this a while back: http://stratus3d.com/blog/2018/08/18/fibonacci-algorithms-in-elixir/. Looks like I need to benchmark the implementation that uses `Stream`.

File: articles/clojure_elixir_fibonacci_mstalker/article.md (hunk 6 above)
Comment (michaelstalker) [reply articles/clojure_elixir_fibonacci_mstalker/article.md:135]: `(inc current-index)` and `(+ current-index 1)` are equivalent. I figured that since Clojure offers an `inc` function, I'd use it, rather than write out the addition.

File: articles/clojure_elixir_fibonacci_mstalker/article.md (hunk 9 above)
Comment (michaelstalker) [reply articles/clojure_elixir_fibonacci_mstalker/article.md:21]: Nice!

File: articles/clojure_elixir_fibonacci_mstalker/article.md (hunk 4 above)
Comment (michaelstalker) [reply articles/clojure_elixir_fibonacci_mstalker/article.md:117]: I'll reword this to "Try calling these functions with an argument of `200000` and you'll see what I mean!"

File: articles/clojure_elixir_fibonacci_mstalker/article.md (hunk 3 above)
Comment (michaelstalker) [reply articles/clojure_elixir_fibonacci_mstalker/article.md:81]: Clojure supports function overloading. I'm not sure if what Elixir does is technically function overloading or not. I like the idea of exploring overloading, but I think I'm going to omit mentioning it here.

File: articles/clojure_elixir_fibonacci_mstalker/article.md (hunk 2 above)
Comment (michaelstalker) [reply articles/clojure_elixir_fibonacci_mstalker/article.md]: Thanks for pointing out a confusing part! I moved some of the explanation to the code comment itself.

I didn't see any Elixir REPL, and the Clojure ones I saw were so-so. Unfortunately, Medium doesn't let me embed a REPL.

File: articles/clojure_elixir_fibonacci_mstalker/article.md (hunk 1 above)
Comment (michaelstalker) [reply articles/clojure_elixir_fibonacci_mstalker/article.md:71]: I dropped the inequality.

**Comment by michaelstalker** [comment]
@mmaiza To keep the code simple, since this is just for teaching purposes, I'm going to omit the guards.

Thanks, everyone, for reviewing this!
